### PR TITLE
Add Go CLI to OpenWhisk while leaving Python CLI functionality intact

### PR DESCRIPTION
**Pull request by dubeejw**
- Add Go CLI source code to OpenWhisk while leaving Python CLI functionality intact

### CODE REVIEW

**Comment by lehoanganh** [comment]
there are some IDE config files, it's needed to push them?
do you plan to move the work to ansible?

**Comment by dubeejw** [comment]
@lehoanganh, the plan is to first add the Go CLI in a way that it does not affect the existing CLI. That's what this pull request accomplishes. See #238 for the Ansible work.

**Comment by csantanapr** [comment]
@dubeejw Like @lehoanganh said please update top level [.gitignore](https://github.com/openwhisk/openwhisk/blob/master/.gitignore)
to ignore all the `.idea/` files in this PR

**Comment by dubeejw** [comment]
@csantanapr, @rabbah, @markusthoemmes, @psuter, anyone opposed to this pull request being merged?

**Comment by rabbah** [comment]
Wait for two thumbs up please.

File: tools/go-cli/go-whisk-cli/Godeps/Readme [hunk 1]
```diff
@@ -0,0 +1,5 @@
+This directory tree is generated automatically by godep.
```
Comment (psuter) [inline tools/go-cli/go-whisk-cli/Godeps/Readme]: Do we need this checked in if it's generated automatically?

File: tools/go-cli/go-whisk-cli/commands/util.go [hunk 2]
```diff
@@ -0,0 +1,583 @@
...
+// 		return nil, err
+// 	}
+// 	return annotations, nil
+// }
+
+func logoText() string {
+
+    logo := `
+
+__          ___     _     _
```
Comment (psuter) [inline tools/go-cli/go-whisk-cli/commands/util.go]: Any we can fit the logo in ASCII art? :)

File: tools/go-cli/go-whisk-cli/notes.md [hunk 3]
```diff
@@ -0,0 +1,736 @@
+
```
Comment (psuter) [inline tools/go-cli/go-whisk-cli/notes.md]: Does this need to be checked in? It could be in an issue.

Comment (sjfink) [inline tools/go-cli/go-whisk-cli/notes.md]: Do you intend to check in these notes?

File: tools/go-cli/go-whisk/Godeps/Godeps.json [hunk 4]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (psuter) [inline tools/go-cli/go-whisk/Godeps/Godeps.json]: Same question as for ` tools/go-cli/go-whisk-cli/Godeps/`

**Comment by sjfink** [comment]
Once this is checked in, can we create a new jenkins build to test it easily?

**Comment by psuter** [comment]
Can we please add this at the end of `tools/build/scanCode.py` and make sure it's enforced:

    ("*.go",       [ has_block_copyright, line_checks([ no_tabs, no_trailing_spaces, eol_at_eof ]) ])

(...and adapt if necessary, e.g. currently the license header uses a slightly different format than the Java/Scala one.)

**Comment by dubeejw** [comment]
@sjfink, this pull request is just for the Go CLI source code, it does not modify the build process in anyway as we want to get some source code in ASAP. However, I have another pull request open just for the purpose of running the Go CLI against all the automation in the TravisCI build. I am currently working on getting the Jenkins builds to pass using the Go CLI.

**Comment by csantanapr** [comment]
I sat with James I did a review, here are my findings.

Things to fix in the PR to merge it in:

Add Readme.md to go-cli/README.md
Remove "BlueMix" from go-cli/go-whisk-cli/Godeps
Delete go-cli/go-whisk-cli/notes.md
Remove everything except `Godeps/_worspace` from go-cli/.gitignore

Add to the README.md the use of the env variable WSK_CLI_DEBUG

Move properties to config.xml
    <property name="go-cli.build.dir"        value="${basedir}/build/tools/go-cli" />
    <property name="go-cli.src.dir"          value="${basedir}/tools/go-cli" />

Action items outside this PR:
- look into depenencies versions we are currently using (go-whiks-cli/Godeps/Godeps.json)
- look into GoVersion declare (1.5)go-whiks-cli/Godeps/Godeps.json)
- `wsk property get` doesn't return info like python cli, need to pass --all
- wsk cli apihost is hardcoded to (openwhisk.ng.bluemix.net), and it should be set to the value of edge.host from whisk.properties (i.e. edge.host=172.17.0.1)
- the order of the properties don't mathc the python cli output with `wsk property get`
- Positional parameters are not documented like python cli
wsk action create <name string> <artifact string> [flags]
the should be like this or similar: human description for the paramters
positional arguments:
  name                  the name of the action
  artifact              artifact (e.g., file name) containing action
                        definition
- Need to print a new line after ok result when creating a new action
$ ./wsk action create myaction myaction.js
ok: created action myactionvagrant@vagrant-ubuntu-trusty-64:~/openwhisk/bin/go-cli/linux/amd64$

- can't pass numbers as parameters 
./wsk -v action invoke myaction --blocking --result -p foo [1,2,3]
They get pass as string array instead of number array
Req Body
{{"foo":["1","2","3"]}
}

In the python CLI it get's pass as numbers
Body sent:
{"foo": [1, 2, 3]}
In python to pass strings you need to escape the quotes
 ./wsk -v action invoke myaction --blocking --result -p foo [\"1\",\"2\",\"3\"]
 Body sent:
{"foo": ["1", "2", "3"]}


Tips to improve later:
- make env variables to only build one binary making dev cycle faster
- look into implementing a watch/make task to automatically compile on source code change 
- upgrade version of from 1.6 to 1.7, 1.7 is GA
- implement autocomplete for CLI ( we have recommendations when typos, but autocomplete will be also good to have)s

**Comment by dubeejw** [comment]
@psuter, @csantanapr, I think I got all of your suggested changes.

**Comment by csantanapr** [comment]
@dubeejw Could you open issues for the work items I list in section 2 (2. Action items outside this PR:) in my comment above https://github.com/openwhisk/openwhisk/pull/604#issuecomment-224987914

currently trying out your recent changes

**Comment by csantanapr** [comment]
I did a license check on the dependencies, and only one raises a concern: - http://github.com/go-yaml/yaml/tree/v2 the license is LGPLv3

@rabbah @dubeejw Should we resolve this before we merge this code in?

- http://github.com/mattn/go-isatty

MIT
- http://github.com/fatih/color

MIT
- http://github.com/mattn/go-colorable

MIT
- http://github.com/mitchellh/go-homedir

MIT
- http://github.com/google/go-querystring/

BSD-style

- http://github.com/hokaccha/go-prettyjson

MIT
- http://github.com/spf13/pflag

BSD-style

- http://github.com/spf13/cobra

Apache 2.0
- http://github.com/spf13/viper

MIT
- http://github.com/inconshreveable/mousetrap

Apache 2.0
- http://github.com/cpuguy83/go-md2man

MIT
- http://github.com/stretchr/testify/

MIT
- http://github.com/fsnotify/fsnotify

MIT
- http://github.com/mitchellh/mapstructure

MIT
- http://github.com/spf13/cast

MIT
- http://github.com/spf13/jwalterweatherman

MIT
- http://github.com/BurntSushi/toml

DO WHAT THE F*** YOU WANT TO PUBLIC LICENSE
- http://github.com/hashicorp/hcl

Mozilla Public License, version 2.0

- http://github.com/magiconair/properties

MIT
- http://github.com/go-check/check/tree/v1

MIT
- http://github.com/go-yaml/yaml/tree/v2

LGPLv3

- http://github.com/xordataexchange/crypt/

MIT
- http://github.com/russross/blackfriday

Simplified BSD

- http://github.com/stretchr/objx

MIT
- http://github.com/davecgh/go-spew/

MIT
- http://github.com/pmezard/go-difflib/

MIT
- http://github.com/ernesto-jimenez/gogen/

MIT
- http://github.com/coreos/go-etcd/

Apache 2.0
- http://github.com/armon/consul-api

Mozilla Public License, version 2.0

- http://github.com/shurcooL/sanitized_anchor_name

MIT
- http://github.com/ugorji/go/

MIT

**Comment by csantanapr** [comment]
I think is safe to take the yaml (Go) as dependency, they have an exception in their [ yaml (Go) License](https://github.com/go-yaml/yaml/blob/v2/LICENSE)

> As a special exception to the GNU Lesser General Public License version 3
("LGPL3"), the copyright holders of this Library give you permission to
convey to a third party a Combined Work that links statically or dynamically
to this Library without providing any Minimal Corresponding Source or
Minimal Application Code as set out in 4d or providing the installation
information set out in section 4e,...

I give a 👍 to merge this in.

I added new TODO in my comment above https://github.com/openwhisk/openwhisk/pull/604#issuecomment-224987914:
>- Pinned the versions of the dependencies, we should not have a dependency again the head/master of the git repo, relying on their repos to be available, and assume they don't do something that will break our CLI is a recipe for disaster. We can always update and refresh the dependency on a regular cadence.

File: tools/go-cli/go-whisk/whisk/activation.go [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+        errStr := fmt.Sprintf("Request failure: %s", err)
+        werr := MakeWskErrorFromWskError(errors.New(errStr), err, EXITCODE_ERR_NETWORK, DISPLAY_MSG, NO_DISPLAY_USAGE)
+        return nil, resp, werr
+    }
+
+    return activation, resp, nil
+}
+
+func (s *ActivationService) Result(activationID string) (*Response, *http.Response, error) {
+    // TODO :: for some reason /activations only works with "_" as namespace
```
Comment (rabbah) [inline tools/go-cli/go-whisk/whisk/activation.go]: All activations are stored in the default namespace. Using another namespaces will result in an empty response `[]`. The reason is that activating a trigger/rule/action cannot distinguish the context (namespace) within which to record the activation record.

**Comment by rabbah** [comment]
I expect the code will be improved further on subsequent iterations - in the interest of expedience to get to the next step 👍.

**Comment by csantanapr** [comment]
rebase and merged bce6eb86a3cde3ab12354ea158a0ce1708ff993a

**Pull request closed without merging**